src/lib.rs

```diff
@@ -273,12 +276,29 @@ where
     Ok(state)
 }
 
+#[derive(Debug)]
+pub enum CmarkError {
+    FormatFailed(fmt::Error),
+    UnexpectedEvent,
+}
+
+impl fmt::Display for CmarkError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::FormatFailed(e) => e.fmt(f),
+            Self::UnexpectedEvent => f.write_str("Unexpected event while reconstructing Markdown"),
+        }
+    }
+}
+
+impl std::error::Error for CmarkError {}
```

Comment:
I think this should implement `From<fmt::Error>`, so the `write_str` function can be removed.

Reply:
I'm still new to Rust so I'm likely failing to notice how implementing `From<fmt::Error>` can shortcut `write_str`. Would be nice if this provided an automatic implementation of `From<fmt::Result>` for `Result<(), Error>`, so that then I could simply `.into()` every time, but that seems not to be the case... :face_with_diagonal_mouth:

Could you give me a more explicit hint? Thanks!

Reply:
It's implemented for `Error`, the `?` will do the right thing automatically.

Reply:
Ah okay, didn't know that! I had to wrap the last `match` statement in an `Ok(match { ... }?)` to make it work, but otherwise I reverted all `write_char` and `write_str` to the original.